File: LeetCode/Medium/1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.py
class Solution:
    def maxSatisfied(self, customers: List[int], grumpy: List[int], minutes: int) -> int:
        n = len(customers)
        ret = 0
        unsatisfied = 0
        

        for i in range(minutes):
            if grumpy[i]:
                unsatisfied += customers[i]
            else:
                ret += customers[i]

        for i in range(minutes, n):
            if not grumpy[i]:
                ret += customers[i]
        
        max_val = unsatisfied

        for i in range(minutes, n):
            if grumpy[i-minutes]:
                unsatisfied -= customers[i-minutes]
            if grumpy[i]:
                unsatisfied += customers[i]
            max_val = max(max_val, unsatisfied)

        return ret + max_val

File: LeetCode/Medium/1052-grumpy-bookstore-owner/1052-grumpy-bookstore-owner.go
func maxSatisfied(customers []int, grumpy []int, minutes int) int {
    n := len(customers)
    ret := 0
    unsatisfied := 0

    for i := 0; i < minutes; i++ {
        if grumpy[i] == 1 { 
            unsatisfied += customers[i] 
        } else { ret += customers[i] }
    }

    for i := minutes; i < n; i++ {
        if grumpy[i] == 0 { ret += customers[i] }
    }

    maxVal := unsatisfied

    for i := minutes; i < n; i++ {
        if grumpy[i-minutes] == 1 { unsatisfied -= customers[i-minutes] }
        if grumpy[i] == 1 { unsatisfied += customers[i] }
        maxVal = max(maxVal, unsatisfied)
    }

    return ret + maxVal
}


func max(a, b int) int {
    if a > b { return a }
    return b
}